### DISCUSSION TITLE
Pin state dependent D0-SLM counter selection

### DISCUSSION THREAD

**Discussion by HaraldNoack** [arendst/Tasmota#17546] [Q&A] [answered]
Hello,

is it possible to write a script that uses the S0-SLM-Counter1 when the GPIO13 pin is LOW and increment  S0-SLM-Counter2 when the GPIO13 pin is HIGH ?
The S0-impuls-input is GPIO12.

This should be the solution for the following problem:
I have a heat pump and I want to measure the energy that is used for heating or hot water production.
I have a S0-energymeter with 1000pulse/kWh connected to the heat pump and a reed-switch with magnet connected to the 3-way-valve. => I want to count the energy valve-state-dependent. 
Energy for heating
Energy for hot water production

I defined 2 counter in this script:
`>D 
>B
>M 2
+1,12,c,1,-10,Heizung
1,1-0:1.8.0*255(@1000, Energie,KWh,Total_in,3)
1,1-0:1.7.0*255(@0.01667, Leistung,W,Power_actual,0)
+2,14,c,1,-10,WWasser
2,1-0:1.8.0*255(@1000, Energie,KWh,Total_in,3)
2,1-0:1.7.0*255(@0.01667, Leistung,W,Power_actual,0)
#`
How have I to modify it to get the desired behaviour?

Thanks for your HELP!!!!!
Best Regards
Harald

**Comment by gemu2015** [comment]
this will not work, you can not switch the meter input dependent on a pin
define a single counter meter and depending on the pin state changes of your reed contact add the counter reading delta to a variable. this requires a bit of knowledge in scripting.
since this is an interesting use case i will help with the script when i find some time.

**Reply by HaraldNoack** [reply]
Thanks for your idea !!!

My first idea was to define 2 counter in a script, as in the question, connected to 2 GPIO-Pins and switch the pulses from the S0-contact controlled by the reed switch between the 2 GPIO-Pins.

The problem with this solution is, that wenn no new impulse is detected by a counter-pin, then the old "actual power" "1-0:1.7.0255" stands constant on a level forever. That means, the old outdated power is displayed.

Maybe someone know a solution for this?

It is also possible to define a 3. S0-counter with GPIO that receives all pulses and displays the correct "actual power" and that value is copied to the corresponding "1-0:1.7.0255" of the other counter. I don't know if this is possible ?

I thought maybe there is an easy, only software solution without making a small pcb.

TAHNKS for your HELP!!!!
Best Regards
Harald

**Comment by gemu2015** [comment, marked as answer]
a very quick start to begin with

```
>D
p:water=0
water_last=0
p:heater=0
heater_last=0
valve=0
>B
->sensor53 r
; set pinmode of reed contact to input pullup
spinm(13 P)
is(0 "Heizung|Wasser|")
>F
; read pin
valve=pin[13]
if chg[valve]>0 {
; state has changed
	if valve==0 {
	; from high to low
		heater_last=sml[1]
	} else {
	; from low to high
		water_last=sml[1]
	}
}

if valve>0 {
	// water
	water+=sml[1]-water_last
	water_last=sml[1]
} else {
	// heater
	heater+=sml[1]-heater_last
	heater_last=sml[1]
}


>M 1
+1,12,c,1,-10,Heizung
1,1-0:1.8.0255(@1000, Energie,KWh,Total_in,3)
1,1-0:1.7.0255(@0.01667, Leistung,W,Power_actual,0)
#

>W
Wasser:{m}%water% kWh
Heizung:{m}%heater% kWh
Status:{m}%is[valve+1]%
```

**Reply by HaraldNoack** [reply]
THANKS!!!!

That looks great and I have to test it the next days :-)

**Reply by gemu2015** [reply]
the above proposal (edited) only deals with kWh not with current consumption. you have to expand that.
however measuring with pulse counters has its problems as they are updated only with a dedicated pulse.
i strongly would prefer other readings as serial OBIS,SML or MODBUS.
